client_plugin/drivers/shared/etcdops.go

```diff
@@ -280,3 +288,122 @@ func addrToEtcdClient(addr string) (*etcdClient.Client, error) {
 
 	return cli, nil
 }
+
+// Update or Create volume metadata in KV store
+func (d *VolumeDriver) writeVolMetadata(name string,
+	record VolumeMetadata) error {
+
+	var msg string
+	var value string
+	var err error
+
+	value = fmt.Sprintf("%d:", record.port)
+	value += record.username + ":"
+	value += record.password + ":"
+	for _, element := range record.clientList {
+		value += element
+		value += "?"
+	}
+
+	etcdAPI := d.createEtcdClient()
+	defer etcdAPI.Close()
+
+	// Write grefc
+	ctx, cancel := context.WithTimeout(context.Background(),
+		kvStoreRequestTimeout*time.Second)
+	_, err = etcdAPI.Put(ctx, keyPrefix+name+refKey, fmt.Sprintf("%d", record.globalRefcount))
+	cancel()
+	if err != nil {
+		msg = fmt.Sprintf("Failed to write grefc metadata for volume %s. Reason: %v", name, err)
+		log.Warningf(msg)
+		return errors.New(msg)
+	}
+
+	// Write volume status
+	ctx, cancel = context.WithTimeout(context.Background(),
+		kvStoreRequestTimeout*time.Second)
+	_, err = etcdAPI.Put(ctx, keyPrefix+name+statusKey, record.status)
```

Comment:
does the API have "bulk put"  or transactions ? There are too many race opportunities here. 
Maybe you should have a lock (via etcd KV pair) and have others wait on the lock release ?